### DISCUSSION TITLE
Mypy doesn't respect validators

### DISCUSSION THREAD

**Discussion by stinovlas** [pydantic/pydantic#5147] [Question] [answered]
I'm not sure whether this is a problem of pydantic mypy plugin or not.

Consider following code:

```python
from datetime import datetime
from pydantic import BaseModel

class A(BaseModel):
    d: datetime

print(A(d="2022-01-01T00:00:00Z"))
```

This code works without problems and `pydantic` flawlessly parses string to correct datetime. However, when I run `mypy==1.1.1` with `plugins = pydantic.mypy`, I get following error:

```
test.py:7: error: Argument "d" to "A" has incompatible type "str"; expected "datetime"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

On the other hand, with `mypy==1.0.1`, there is no error. Since `pydantic` is able to accept string and turn it into `datetime` using validators, I don't think that mypy should complain about this. But it does.

I chose `datetime` because it's part of the standard library, but I also have my own field types inheriting from `str` and mypy complains about those as well. Is this a problem that `pydantic.mypy` plugin should resolve?

I found https://github.com/pydantic/pydantic/issues/4680 which indicates this might be covered by `pydantic v2`, but until then, what's the recommended course of action?

**Comment by dmontagu** [comment, marked as answer]
This is because mypy now respects the @dataclasses.dataclass_transform decorator in 1.1.0. There is some open work to resolve this that I suspect will be merged very shortly. I'll make a note to message here once that is released.

**Reply by dmontagu** [reply]
I just looked into this a bit more, and I think once #5111 gets merged, the plugin should work properly with mypy 1.1.1 when used in "lenient" mode. Note that to prevent errors like those you are indicating above, you'll now _need_ to use the plugin to get "lenient" behavior as, without the plugin, mypy 1.1.1 will be treating subclasses of `BaseModel` in a manner similar to how it treats dataclasses.

I'm hoping that we can get #5111 merged and a new release out tomorrow.

**Reply by stinovlas** [reply]
This is very encouraging, thank you for your time! I'll be looking forward to the next release :-).

**Reply by dmontagu** [reply]
Okay, just created the 1.10.6 release, GitHub Actions should post it to PyPI shortly.

**Reply by stinovlas** [reply]
Thank you very much. I verified that the problem presented in the original post is now resolved.
